// arls-kaiyo-lite/models/response-api.model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseApiError = exports.ResponseApiWarning = exports.ResponseApiSuccess = void 0;
var arls_logger_1 = __importDefault(require("arls-logger"));
var ResponseApiSuccess = /** @class */ (function () {
    function ResponseApiSuccess(result, msg) {
        if (result === void 0) { result = null; }
        if (msg === void 0) { msg = ""; }
        this.success = true;
        this.error = false;
        this.warning = false;
        this.message = "";
        this.result = null;
        this.result = result;
        this.message = msg;
    }
    ResponseApiSuccess.prototype.setPagination = function (pageTotal, itemsTotal, fullItemsTotal) {
        if (itemsTotal === void 0) { itemsTotal = null; }
        if (fullItemsTotal === void 0) { fullItemsTotal = null; }
        this.pageTotal = pageTotal;
        if (itemsTotal)
            this.itemsTotal = itemsTotal;
        if (fullItemsTotal)
            this.fullItemsTotal = fullItemsTotal;
        return this;
    };
    return ResponseApiSuccess;
}());
exports.ResponseApiSuccess = ResponseApiSuccess;
var ResponseApiWarning = /** @class */ (function () {
    function ResponseApiWarning(msg, warningDetails) {
        if (msg === void 0) { msg = ""; }
        if (warningDetails === void 0) { warningDetails = null; }
        this.success = false;
        this.error = false;
        this.warning = true;
        this.message = "";
        this.result = null;
        this.message = msg;
        if (warningDetails) {
            arls_logger_1.default.info(msg, warningDetails);
        }
    }
    return ResponseApiWarning;
}());
exports.ResponseApiWarning = ResponseApiWarning;
var ResponseApiError = /** @class */ (function () {
    function ResponseApiError(msg, errorDetails) {
        if (msg === void 0) { msg = ""; }
        if (errorDetails === void 0) { errorDetails = null; }
        this.success = false;
        this.error = true;
        this.warning = false;
        this.message = "";
        this.result = null;
        this.message = msg;
        if (errorDetails) {
            arls_logger_1.default.error(msg, errorDetails);
        }
    }
    return ResponseApiError;
}());
exports.ResponseApiError = ResponseApiError;
